### ISSUE TITLE
bug: no-derived-state false positive with async function that await a dependency

### ISSUE THREAD

**Issue by TkDodo** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#35]
### Did you check docs and existing issues?

- [x] I have read the plugin docs
- [x] I have searched the existing issues
- [x] I have read [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect) and am reasonably confident that I *do* need an effect

### Plugin version

0.5.1

### ESLint version

9.34.0

### Describe the bug

the following code shouldn’t error, but it does:

```
const api = useFetchWrapper();
const [state, setState] = useState();

useEffect(() => {
  async function fetchIt() {
    const response = await api.doFetch(`/my/url`);
    setState(response);
  }

  void fetchIt();
}, [api]);
```

note that if we call `await fetch()` directly and remove `api` from the dependency array, there is no violation.

### Steps To Reproduce

1. run the code above
2. there should be no lint error

### Expected Behavior

there should be no error.

**Comment by TkDodo** [comment]
interesting that here, only one of the calls to `setNewInsightIndices` errors?

```
const previousInsightsRef = useRef([]);
// Compare current insights with previous insights to determine which ones are new
useEffect(() => {
  if (insights.length === previousInsightsRef.current.length + 1) {
    // eslint-disable-next-line react-you-might-not-need-an-effect/no-derived-state
    setNewInsightIndices([insights.length - 1]);
  } else {
    setNewInsightIndices([]);
  }
  previousInsightsRef.current = [...insights];
}, [insights]);
```
